interop-cats/jvm/src/test/scala/scalaz/zio/interop/catzSpec.scala

```diff
@@ -102,25 +104,44 @@ class catzSpec
 
   // TODO: reintroduce repeated ConcurrentTests as they're removed due to the hanging CI builds (see https://github.com/scalaz/scalaz-zio/pull/482)
   checkAllAsync(s"ConcurrentEffect[Task]", implicit tctx => IOConcurrentEffectTests().concurrentEffect[Int, Int, Int])
-  checkAllAsync("Concurrent[Task]", (_) => ConcurrentTests[Task].concurrent[Int, Int, Int])
   checkAllAsync("Effect[Task]", implicit tctx => EffectTests[Task].effect[Int, Int, Int])
+  checkAllAsync("Concurrent[TaskR[Int, ?]]", (_) => ConcurrentTests[TaskR[Int, ?]].concurrent[Int, Int, Int])
+  checkAllAsync("Concurrent[TaskR[String, ?]]", (_) => ConcurrentTests[TaskR[String, ?]].concurrent[Int, Int, Int])
+  checkAllAsync("Concurrent[Task]", (_) => ConcurrentTests[Task].concurrent[Int, Int, Int])
+  checkAllAsync(
+    "MonadError[ZIO[String, Int, ?]]",
+    (_) => MonadErrorTests[ZIO[String, Int, ?], Int].monadError[Int, Int, Int]
+  )
   checkAllAsync("MonadError[IO[Int, ?]]", (_) => MonadErrorTests[IO[Int, ?], Int].monadError[Int, Int, Int])
+  checkAllAsync(
+    "Alternative[ZIO[String, Int, ?]]",
+    (_) => AlternativeTests[ZIO[String, Int, ?]].alternative[Int, Int, Int]
+  )
   checkAllAsync("Alternative[IO[Int, ?]]", (_) => AlternativeTests[IO[Int, ?]].alternative[Int, Int, Int])
   checkAllAsync(
     "Alternative[IO[Option[Unit], ?]]",
     (_) => AlternativeTests[IO[Option[Unit], ?]].alternative[Int, Int, Int]
   )
-  checkAllAsync("SemigroupK[UIO[?]]", (_) => SemigroupKTests[UIO[?]].semigroupK[Int])
+  checkAllAsync("SemigroupK[TaskR[String, ?]]", (_) => SemigroupKTests[TaskR[String, ?]].semigroupK[Int])
+  checkAllAsync("SemigroupK[Task]", (_) => SemigroupKTests[Task].semigroupK[Int])
+  checkAllAsync(
+    "Bifunctor[ZIO[String, ?, ?]]",
+    (_) => BifunctorTests[ZIO[String, ?, ?]].bifunctor[Int, Int, Int, Int, Int, Int]
+  )
   checkAllAsync("Bifunctor[IO]", (_) => BifunctorTests[IO].bifunctor[Int, Int, Int, Int, Int, Int])
   checkAllAsync("Parallel[Task, Task.Par]", (_) => ParallelTests[Task, Util.Par].parallel[Int, Int])
+  checkAllAsync(
+    "Parallel[TaskR[Int, ?], Task.ParIO[IO, Throwable, ?]]",
+    (_) => ParallelTests[TaskR[Int, ?], ParIO[Int, Throwable, ?]].parallel[Int, Int]
+  )
 
-  implicit def catsEQ[E, A: Eq]: Eq[IO[E, A]] =
-    new Eq[IO[E, A]] {
+  implicit def catsEQ[R, E, A: Eq]: Eq[ZIO[R, E, A]] =
+    new Eq[ZIO[R, E, A]] {
       import scalaz.zio.duration._
 
-      def eqv(io1: IO[E, A], io2: IO[E, A]): Boolean = {
-        val v1  = unsafeRunSync(io1.timeout(20.seconds)).map(_.get)
-        val v2  = unsafeRunSync(io2.timeout(20.seconds)).map(_.get)
+      def eqv(io1: ZIO[R, E, A], io2: ZIO[R, E, A]): Boolean = {
+        val v1  = unsafeRunSync(io1.asInstanceOf[IO[E, A]].timeout(20.seconds)).map(_.get)
```

Comment:
This needs to use `provide`

Reply:
How do I get the R in this case? By creating an `Arbitrary` for `R`?

Reply:
You need an implicit runtime. Or you can provide equality for `IO[E, A]` or even `Task[A]` if the tests will permit it. I don't think the tests use anything except for `Task`.

Reply:
@jdegoes as I mentioned on Gitter I'm still lost with this part. Adding an `implicit rts: Runtime[R]` constraint to derive the `Eq` instances makes sense but the issue is where to get this instance from? If you look at how the spec class is defined maybe you'll see the problem?

```scala
class catzSpec
    extends FunSuite
    with BeforeAndAfterAll
    with Matchers
    with Checkers
    with Discipline
    with TestInstances
    with GenIO
    with DefaultRuntime {

  override val Platform = PlatformLive.makeDefault().withReportFailure(_ => ())

  implicit val runtime: Runtime[Any] = this
```

If I remove the `with DefaultRuntime` how do I get a runtime? This class is instantiated by the test framework AFAIU.

Reply:
@gvolpe you can just do `implicit val runtime: Runtime[Any] = new DefaultRuntime {}`

Reply:
The question is how do I get a `Runtime[R]`? @ghostdogpr

Reply:
Sorry I missed your point. You're only using R = Int and R = String in the tests so isn't it sufficient to declare 2 implicit runtimes, one for Int and one for String with arbitrary values?

Reply:
I think that's not what John has in mind so I would love to know :D . We should be able to have tests for any kind of `R` and writing one for each doesn't seem convenient. I'm also very confused with this because it's exactly the reason why `cats-effect` doesn't provide instances of `ConcurrentEffect` and `Effect` for `Kleisli`. How do you run a function `A => F[B]`? What's the `A`?

Reply:
You need to get the R from somewhere. Define an implicit def that for any type `R: Arbitrary` can return a `Gen[Runtime[R]]`. Then you can easily summon a Runtime for any arbitrary R inside the test.

A more hacky way to do that is to define an implicit def that provides an inolicit Runtime for any Arbitrary R. By running the Gen to get some R.

Reply:
The arbitrary trick is what I suggested at first but we were probably thinking of different things :laughing: Now it makes sense, will try to get it done! Thanks :)

Reply:
@jdegoes Unfortunately I'm still stuck with this. The idea sounds good but it seems I require more than just `Runtime[R]`: I need `Runtime[R with Clock with System .... ]`. So this was my attempt:

```scala
  implicit val anyArb: Arbitrary[Any] = Arbitrary(Gen.const(()))

  type Env[R] = R with Clock with Console with System with Random with Blocking

  implicit def deriveRuntime[R: Arbitrary]: Gen[Runtime[Env[R]]] =
    Arbitrary.arbitrary[R].map { env =>
      Runtime[Env[R]](runtime.Environment + env, PlatformLive.Default)
    }

  implicit def catsEQ[R: Arbitrary, E, A: Eq](implicit rts: Runtime[Env[R]]): Eq[ZIO[R, E, A]] =
    new Eq[ZIO[R, E, A]] {
      import scalaz.zio.duration._

      def eqv(io1: ZIO[R, E, A], io2: ZIO[R, E, A]): Boolean = {
        val v1  = rts.unsafeRunSync(io1.timeout(20.seconds)).map(_.get)
        val v2  = rts.unsafeRunSync(io2.timeout(20.seconds)).map(_.get)
        val res = v1 === v2
        if (!res) {
          println(s"Mismatch: $v1 != $v2")
        }
        res
      }
    }

  implicit def catsParEQ[R: Arbitrary, E: Eq, A: Eq](implicit rts: Runtime[Env[R]]): Eq[ParIO[R, E, A]] =
    new Eq[ParIO[R, E, A]] {
      def eqv(io1: ParIO[R, E, A], io2: ParIO[R, E, A]): Boolean =
        rts.unsafeRun(Par.unwrap(io1).either) === rts.unsafeRun(Par.unwrap(io2).either)
    }
```

In order to create the environment I need to mixin some traits (`Runtime.apply`) and all I've got is a plain value of type `R` so this doesn't seem possible.

Also notice that where I require an `implicit rts: Runtime[Env[R]]` I'm not going to get one. Instead I'll have a `Gen` in scope to create a `Runtime[Env[R]]` which seems overkill. Why not just casting the given `ZIO[R, E, A]` to `IO[E, A]` as I've done to get it working? After all it's just tests to `unsafeRunSync` some computation but if there's a better way I'd like to know it.

So once again I'm on a dead end here :disappointed:

Reply:
Any attempt to cast away the `R` will result in a runtime exception somewhere, unless the `R` is not actually used. 

I see now that you cannot use an arbitrary `R` here. If you like, you could create a "test" module, like this one:

```scala
trait TestModule[+A] {
  def testModule: A
}
```

Then you could make an `Arbitrary` for that (given an `Arbitrary` for `A`), and finally, you could supply an `Arbitrary[Runtime[TestModule[A] with Clock with System ...]`.

However, the question is why? In my opinion, testing with `R = Int` and `R = String` like is being done is not that useful. You may as well fix the environment to `Clock with System` or weaker in the tests, rather than using `R = Int` or `R = String`.

If you fix the environment to the above, you can just use `implicit val TestRuntime = new DefaultRuntime`. Then fix the two or so compiler errors where a test is using a different `R`. Then you should be good to go.

It's a nice idea in theory to make the test `R agnostic` but you can't use a single `R` for the whole test suite in that case, which will cause lots of pain. Just best to fix the `R` for the test suite, because it's already as polymorphic as it can be in the instances.

Reply:
Thanks for bearing with me, John.

> However, the question is why? In my opinion, testing with R = Int and R = String like is being done is not that useful. You may as well fix the environment to Clock with System or weaker in the tests, rather than using R = Int or R = String.

You're right. I just want to verify that there are Cats instances in scope for `TaskR[R, ?]` so I could just add the code that summons such instances somewhere, don't need property tests. If something is wrong, the code just won't compile.

> If you fix the environment to the above, you can just use implicit val TestRuntime = new DefaultRuntime. Then fix the two or so compiler errors where a test is using a different R. Then you should be good to go.

I don't understand this. Having a `DefaultRuntime` works fine for the other tests but what do you mean with "fix the compiler errors where a test is using a different `R`"? I think this is the whole issue and why I'm stuck here, it needs a `Runtime[R]` :smile: 

> It's a nice idea in theory to make the test R agnostic but you can't use a single R for the whole test suite in that case, which will cause lots of pain. Just best to fix the R for the test suite, because it's already as polymorphic as it can be in the instances.

Completely agree :+1: I just want to make this as simple as possible, I just haven't found the way.